apps/emqx_auth/src/emqx_authz/emqx_authz_utils.erl

```diff
@@ -41,50 +43,86 @@
 %% APIs
 %%--------------------------------------------------------------------
 
-create_resource(ResourceId, Module, Config, Type) ->
-    Result = emqx_resource:create_local(
-        ResourceId,
-        ?AUTHZ_RESOURCE_GROUP,
-        Module,
-        Config,
-        ?DEFAULT_RESOURCE_OPTS(Type)
-    ),
-    start_resource_if_enabled(Result, ResourceId, Config).
-
-update_resource(Module, #{annotations := #{id := ResourceId}} = Config, Type) ->
-    Result =
-        case
+-spec create_resource(module(), map()) -> ok | no_return().
+create_resource(
+    ConnectorModule,
+    #{resource_id := ResourceId, type := Type, resource_config := ResourceConfig} = State
+) ->
+    maybe
+        {ok, _} ?=
+            emqx_resource:create_local(
+                ResourceId,
+                ?AUTHZ_RESOURCE_GROUP,
+                ConnectorModule,
+                ResourceConfig,
+                ?DEFAULT_RESOURCE_OPTS(Type)
+            ),
+        ok ?= start_resource_if_enabled(State)
+    else
+        {error, Reason} ->
+            error({create_resource_error, Reason})
+    end.
+
+-spec update_resource(module(), map()) -> ok | no_return().
+update_resource(
+    ConnectorModule,
+    #{resource_id := ResourceId, type := Type, resource_config := ResourceConfig} = State
+) ->
+    maybe
+        {ok, _} ?=
             emqx_resource:recreate_local(
                 ResourceId,
-                Module,
-                Config,
+                ConnectorModule,
+                ResourceConfig,
                 ?DEFAULT_RESOURCE_OPTS(Type)
-            )
-        of
-            {ok, _} -> {ok, ResourceId};
-            {error, Reason} -> {error, Reason}
-        end,
-    start_resource_if_enabled(Result, ResourceId, Config).
+            ),
+        ok = start_resource_if_enabled(State)
+    else
+        {error, Reason} ->
+            error({update_resource_error, Reason})
```

Comment:
Curious: why we allow the creation/update of authn resources that fail to start non-explosively, but here (authz) we opt to raise an error?

Reply:
Here we also allow to fail to start, see `start_resource_if_enabled`.

The difference is how we return error:
* authz provider callback expects just the state, so we raise on error;
* authn provider callback supports `{ok, _}/{error, _}` so we return error.